test/unit/language-extensions/iterable.spec.ts

```diff
@@ -0,0 +1,231 @@
+    util.testFunction`
+        ${testMultiIterable}
+        const results: string[] = [];
+        for (${initializer} of testMultiIterable()) {
+            results.push(x);
+            results.push(y);
+        }
+        return results;
+    `
+        .setOptions(iterableProjectOptions)
+        .expectToEqual(["a1", "a2", "b1", "b2", "c1", "c2"]);
+});
+
+test("LuaIterable<LuaMultiReturn> with external control variables", () => {
+    util.testFunction`
+        ${testMultiIterable}
+        const results: string[] = [];
+        let x: string, y: string;
+        for ([x, y] of testMultiIterable()) {
+            results.push(x);
+            results.push(y);
+        }
+        return results;
+    `
+        .setOptions(iterableProjectOptions)
+        .expectToEqual(["a1", "a2", "b1", "b2", "c1", "c2"]);
+});
+
+test.each([".testIterable", '["testIterable"]'])("LuaIterable property", access => {
+    util.testFunction`
+        ${testIterableProperty}
+        const results: string[] = [];
+        for (const s of tester${access}) {
+            results.push(s);
+        }
+        return results;
+    `
+        .setOptions(iterableProjectOptions)
+        .expectToEqual(["a", "b", "c"]);
+});
+
+function makeForwardTests(call: string, code: string) {
+    return [`${code} function forward() { return ${call}; }`, `${code} const forward = () => ${call};`];
+}
+
+test.each(
+    [
+        ["testIterable()", testIterable],
+        ["tester.testIterable", testIterableProperty],
+    ].map(([call, code]) => makeForwardTests(call, code))
+)("LuaIterable return forward", forwardFunction => {
+    util.testFunction`
+        ${forwardFunction}
+        const results: string[] = [];
+        for (const s of forward()) {
+            results.push(s);
+        }
+        return results;
+    `
+        .setOptions(iterableProjectOptions)
+        .expectToEqual(["a", "b", "c"]);
+});
+
+test.each(makeForwardTests("testMultiIterable()", testMultiIterable))(
+    "LuaIterable<LuaMultiReturn> return forward",
+    forwardFunction => {
+        util.testFunction`
+        ${forwardFunction}
+        const results: string[] = [];
+        for (const [x, y] of forward()) {
+            results.push(x);
+            results.push(y);
+        }
+        return results;
+    `
+            .setOptions(iterableProjectOptions)
+            .expectToEqual(["a1", "a2", "b1", "b2", "c1", "c2"]);
+    }
+);
+
+test.each(
+    [
+        ["testIterable()", testIterable],
+        ["testMultiIterable()", testMultiIterable],
+        ["tester.testIterable", testIterableProperty],
+    ].flatMap(
+        ([call, code]): Array<[string, string]> => [
+            [`for (const s in ${call}) {}`, code],
+            [`const i = ${call};`, code],
+            [`function foo(i: any) {} foo(${call});`, code],
+        ]
+    )
+)("invalid use of LuaIterable (%p)", (statement, code) => {
+    util.testFunction`
+        ${code}
+        ${statement}
```

Comment:
It is so hard to understand what this test does, wouldn't it be easier to just declare a simple iterator function and call it here instead of the current 3-fold combinations, maybe splitting this up into 3 different tests for the different iterable types.